nabs/optimize.py

```diff
@@ -0,0 +1,289 @@
+
+
+def optimize(signal, motor, tolerance,
+             average=None, limits=None, method='golden',
+             maximize=False, md=None):
+    """
+    The optimization methods within ``nabs`` take a similar approach to the
+    methodology behind ``scipy.optimize``. Different use cases will have
+    different requirements and it is impossible to devise a single plan that
+    will be optimal for all of them. Instead, ``optimize`` supports different
+    "methods" each with their own strengths and weaknesses. Select between
+    these with the ``method`` parameter. Keep in mind that different
+    methodologies will interpret certain keywords differently. For instance,
+    the ``tolerance`` of the optimization may be interpreted in ways specific
+    to the algorithm chosen. For more information, see the docstring for the
+    specific method.
+
+    Parameters
+    ----------
+    signal: ophyd.Signal
+        Signal to maximize
+
+    motor: ophyd.OphydObject
+        Any set-able object
+
+    tolerance: float, optional
+        The tolerance in which our motor is
+
+    average: int, optional
+        Choice to take an average of points at each point in the scan
+
+    limits: tuple, optional
+        Limit the region the scan will search within. If this is not provided,
+        the soft limits of the signal will be used. In this case, these must be
+        configured or the scan will not be allowed to continue.
+
+    method: str, optional
+        Choice of optimization methods
+
+    maximize: bool, optional
+        Whether to maximize or minimize the signal
+
+    md: dict, optional
+        metadata
+    """
+    # Decide the limits
+    if not limits:
+        # Use the motor limits
+        if hasattr(motor, 'limits') and any(motor.limits):
+            logger.warning("No limits provided. "
+                           "Using the motor soft limits %r",
+                           motor.limits)
+            limits = motor.limits
+        # No provided limits or motor limits. Not allowed.
+        else:
+            raise ValueError("No limits provided or set on motor")
+
+    # Create an inverted signal if we need to maximize
+    if maximize:
+        raw = signal
+        signal = InvertedSignal(raw)
+
+    # Create plan metadata
+    _md = {'detectors': [signal],
+           'motors': [motor],
+           'plan_args': {'signal': repr(signal),
+                         'motor': repr(motor),
+                         'tolerance': tolerance,
+                         'average': average,
+                         'limits': limits,
+                         'method': method,
+                         'maximize': maximize},
+           'plan_name': 'optimize',
+           'hints': {}}
+    try:
+        dimensions = [(motor.hints['fields'], 'primary')]
+    except (AttributeError, KeyError):
+        pass
+    else:
+        _md['hints'].setdefault('dimensions', dimensions)
```

Comment:
Can you explain what this dimensions field is?

Reply:
This informs `BestEffortCallback` what the "independent" variables are.

https://github.com/NSLS-II/bluesky/blob/d9abfdd6ac1d2d9633503d6ec28aa3b267b7e349/bluesky/callbacks/best_effort.py#L95